File: app/src/main/java/com/wireguard/android/activity/MainActivity.java
/*
 * Copyright © 2017-2019 WireGuard LLC. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

package com.wireguard.android.activity;

import android.annotation.SuppressLint;
import android.content.Intent;
import android.os.Bundle;
import androidx.annotation.Nullable;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentTransaction;
import androidx.appcompat.app.ActionBar;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.LinearLayout;

import com.wireguard.android.R;
import com.wireguard.android.fragment.TunnelDetailFragment;
import com.wireguard.android.fragment.TunnelEditorFragment;
import com.wireguard.android.fragment.TunnelListFragment;
import com.wireguard.android.model.Tunnel;

/**
 * CRUD interface for WireGuard tunnels. This activity serves as the main entry point to the
 * WireGuard application, and contains several fragments for listing, viewing details of, and
 * editing the configuration and interface state of WireGuard tunnels.
 */

public class MainActivity extends BaseActivity
        implements FragmentManager.OnBackStackChangedListener {
    @Nullable private ActionBar actionBar;
    private boolean isTwoPaneLayout;
    @Nullable private TunnelListFragment listFragment;

    @Override
    public void onBackPressed() {
        final int backStackEntries = getSupportFragmentManager().getBackStackEntryCount();
        // If the action menu is visible and expanded, collapse it instead of navigating back.
        if (isTwoPaneLayout || backStackEntries == 0) {
            if (listFragment != null && listFragment.collapseActionMenu())
                return;
        }
        // If the two-pane layout does not have an editor open, going back should exit the app.
        if (isTwoPaneLayout && backStackEntries <= 1) {
            finish();
            return;
        }
        // Deselect the current tunnel on navigating back from the detail pane to the one-pane list.
        if (!isTwoPaneLayout && backStackEntries == 1) {
            setSelectedTunnel(null);
            return;
        }
        super.onBackPressed();
    }

    @Override public void onBackStackChanged() {
        if (actionBar == null)
            return;
        // Do not show the home menu when the two-pane layout is at the detail view (see above).
        final int backStackEntries = getSupportFragmentManager().getBackStackEntryCount();
        final int minBackStackEntries = isTwoPaneLayout ? 2 : 1;
        actionBar.setDisplayHomeAsUpEnabled(backStackEntries >= minBackStackEntries);
    }

    // We use onTouchListener here to avoid the UI click sound, hence
    // calling View#performClick defeats the purpose of it.
    @SuppressLint("ClickableViewAccessibility")
    @Override
    protected void onCreate(@Nullable final Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.main_activity);
        actionBar = getSupportActionBar();
        isTwoPaneLayout = findViewById(R.id.master_detail_wrapper) instanceof LinearLayout;
        listFragment = (TunnelListFragment) getSupportFragmentManager().findFragmentByTag("LIST");
        getSupportFragmentManager().addOnBackStackChangedListener(this);
        onBackStackChanged();
        final View actionBarView = findViewById(R.id.action_bar);
        if (actionBarView != null)
            actionBarView.setOnTouchListener((v, e) -> listFragment != null && listFragment.collapseActionMenu());
    }

    @Override
    public boolean onCreateOptionsMenu(final Menu menu) {
        getMenuInflater().inflate(R.menu.main_activity, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(final MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                // The back arrow in the action bar should act the same as the back button.
                onBackPressed();
                return true;
            case R.id.menu_action_edit:
                getSupportFragmentManager().beginTransaction()
                        .replace(R.id.detail_container, new TunnelEditorFragment())
                        .setTransition(FragmentTransaction.TRANSIT_FRAGMENT_FADE)
                        .addToBackStack(null)
                        .commit();
                return true;
            case R.id.menu_action_save:
                // This menu item is handled by the editor fragment.
                return false;
            case R.id.menu_settings:
                startActivity(new Intent(this, SettingsActivity.class));
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    @Override
    protected void onSelectedTunnelChanged(@Nullable final Tunnel oldTunnel,
                                           @Nullable final Tunnel newTunnel) {
        final FragmentManager fragmentManager = getSupportFragmentManager();
        final int backStackEntries = fragmentManager.getBackStackEntryCount();
        if (newTunnel == null) {
            // Clear everything off the back stack (all editors and detail fragments).
            fragmentManager.popBackStackImmediate(0, FragmentManager.POP_BACK_STACK_INCLUSIVE);
            return;
        }
        if (backStackEntries == 2) {
            // Pop the editor off the back stack to reveal the detail fragment. Use the immediate
            // method to avoid the editor picking up the new tunnel while it is still visible.
            fragmentManager.popBackStackImmediate();
        } else if (backStackEntries == 0) {
            // Create and show a new detail fragment.
            fragmentManager.beginTransaction()
                    .add(R.id.detail_container, new TunnelDetailFragment())
                    .setTransition(FragmentTransaction.TRANSIT_FRAGMENT_FADE)
                    .addToBackStack(null)
                    .commit();
        }
    }
}

File: ui/src/main/java/com/jimberisolation/android/activity/MainActivity.kt
/*
 * Copyright © 2017-2023 WireGuard LLC. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
package com.jimberisolation.android.activity

import android.os.Bundle
import android.content.Intent
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.ProgressBar
import androidx.activity.OnBackPressedCallback
import androidx.activity.addCallback
import androidx.appcompat.app.ActionBar
import androidx.fragment.app.FragmentContainerView
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentTransaction
import androidx.fragment.app.commit
import androidx.lifecycle.lifecycleScope
import com.ionspin.kotlin.crypto.LibsodiumInitializer
import com.jimberisolation.android.Application
import com.jimberisolation.android.Application.Companion.getTunnelManager
import com.jimberisolation.android.R
import com.jimberisolation.android.fragment.TunnelDetailFragment
import com.jimberisolation.android.fragment.TunnelEditorFragment
import com.jimberisolation.android.fragment.TunnelListFragment
import com.jimberisolation.android.model.ObservableTunnel
import com.jimberisolation.android.model.TunnelManager
import com.jimberisolation.android.storage.SharedStorage
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

/**
 * CRUD interface for WireGuard tunnels. This activity serves as the main entry point to the
 * WireGuard application, and contains several fragments for listing, viewing details of, and
 * editing the configuration and interface state of WireGuard tunnels.
 */
class MainActivity : BaseActivity(), FragmentManager.OnBackStackChangedListener {
    private var actionBar: ActionBar? = null
    private var isTwoPaneLayout = false
    private var backPressedCallback: OnBackPressedCallback? = null

    private lateinit var spinner: ProgressBar
    private lateinit var detailContainer: FragmentContainerView

    private fun handleBackPressed() {
        val backStackEntries = supportFragmentManager.backStackEntryCount
        // If the two-pane layout does not have an editor open, going back should exit the app.
        if (isTwoPaneLayout && backStackEntries <= 1) {
            finish()
            return
        }

        if (backStackEntries >= 1)
            supportFragmentManager.popBackStack()

        // Deselect the current tunnel on navigating back from the detail pane to the one-pane list.
        if (backStackEntries == 1)
            selectedTunnel = null
    }

    override fun onBackStackChanged() {
        val backStackEntries = supportFragmentManager.backStackEntryCount
        backPressedCallback?.isEnabled = backStackEntries >= 1
        if (actionBar == null) return
        // Do not show the home menu when the two-pane layout is at the detail view (see above).
        val minBackStackEntries = if (isTwoPaneLayout) 2 else 1
        actionBar!!.setDisplayHomeAsUpEnabled(backStackEntries >= minBackStackEntries)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.main_activity)

        var isInitialized = false;
        LibsodiumInitializer.initializeWithCallback {
            isInitialized = true;
        }

        while(!isInitialized) { }

        spinner = findViewById(R.id.loading_spinner)
        detailContainer = findViewById(R.id.detail_container)

        actionBar = supportActionBar

        toggleLoading(true);

        SharedStorage.initialize(this)

        supportActionBar?.displayOptions = ActionBar.DISPLAY_SHOW_CUSTOM;
        supportActionBar?.setCustomView(R.layout.jimber_action_bar);

        isTwoPaneLayout = findViewById<View?>(R.id.master_detail_wrapper) != null
        supportFragmentManager.addOnBackStackChangedListener(this)
        backPressedCallback = onBackPressedDispatcher.addCallback(this) { handleBackPressed() }
        onBackStackChanged()

        lifecycleScope.launch {
            startScreen()
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.main_activity, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            android.R.id.home -> {
                // The back arrow in the action bar should act the same as the back button.
                onBackPressedDispatcher.onBackPressed()
                true
            }

            R.id.menu_action_edit -> {
                supportFragmentManager.commit {
                    replace(R.id.detail_container, TunnelEditorFragment())
                    setTransition(FragmentTransaction.TRANSIT_FRAGMENT_FADE)
                    addToBackStack(null)
                }
                true
            }
            // This menu item is handled by the editor fragment.
            R.id.menu_action_save -> false
            R.id.menu_settings -> {
                startActivity(Intent(this, SettingsActivity::class.java))
                true
            }

            else -> super.onOptionsItemSelected(item)
        }
    }


    override fun onSelectedTunnelChanged(
        oldTunnel: ObservableTunnel?,
        newTunnel: ObservableTunnel?
    ): Boolean {
        val fragmentManager = supportFragmentManager
        if (fragmentManager.isStateSaved) {
            return false
        }

        val backStackEntries = fragmentManager.backStackEntryCount
        if (newTunnel == null) {
            // Clear everything off the back stack (all editors and detail fragments).
            fragmentManager.popBackStackImmediate(0, FragmentManager.POP_BACK_STACK_INCLUSIVE)
            return true
        }
        if (backStackEntries == 2) {
            // Pop the editor off the back stack to reveal the detail fragment. Use the immediate
            // method to avoid the editor picking up the new tunnel while it is still visible.
            fragmentManager.popBackStackImmediate()
        } else if (backStackEntries == 0) {
            // Create and show a new detail fragment.
            fragmentManager.commit {
                add(R.id.detail_container, TunnelDetailFragment())
                setTransition(FragmentTransaction.TRANSIT_FRAGMENT_FADE)
                addToBackStack(null)
            }
        }
        return true
    }

    private suspend fun startScreen() {
        toggleLoading(false)

        val userId = SharedStorage.getInstance().getCurrentUser()?.id
        if (userId == null) {
            withContext(Dispatchers.Main) {
                startActivity(Intent(this@MainActivity, SignInActivity::class.java))
                finish()
            }
            return
        }

        val tunnelManager = getTunnelManager();
        val tunnels = tunnelManager.getTunnelsOfUser();

        if (tunnels.isEmpty()) {
            withContext(Dispatchers.Main) {
                startActivity(Intent(this@MainActivity, SignInActivity::class.java))
                finish()
            }
            return
        }

        withContext(Dispatchers.Main) {
            supportFragmentManager.commit {
                replace(R.id.detail_container, TunnelListFragment())
            }
        }
    }

    private fun toggleLoading(isLoading: Boolean) {
        if (isLoading) {
            spinner.visibility = View.VISIBLE
            detailContainer.visibility = View.GONE
        } else {
            spinner.visibility = View.GONE
            detailContainer.visibility = View.VISIBLE
        }
    }
}
